From 60eaaac2dc6644ed91650f9d7a731563c2c2248f Mon Sep 17 00:00:00 2001
From: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Date: Fri, 19 Jan 2018 09:49:07 +0200
Subject: [PATCH 3233/5242] MLK-17401: drm: imx: dcss: fix scaling issue

commit  e0ebf758ccc4ae8e2bd323c294dc667b92104be4 from
https://source.codeaurora.org/external/imx/linux-imx.git

Under certain circumstances (corner cases) the scaling fractions were
not set properly. This patch fixes this.

Signed-off-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dcss/dcss-scaler.c |   12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/imx/dcss/dcss-scaler.c b/drivers/gpu/imx/dcss/dcss-scaler.c
index 6b85e55..4905ad1 100644
--- a/drivers/gpu/imx/dcss/dcss-scaler.c
+++ b/drivers/gpu/imx/dcss/dcss-scaler.c
@@ -339,10 +339,10 @@ static bool dcss_scaler_fractions_set(struct dcss_soc *dcss, int ch_num,
 {
 	u32 l_vinc, l_hinc, c_vinc, c_hinc;
 
-	l_vinc = (src_yres << 13) / dst_yres;
-	c_vinc = (src_yres << 13) / dst_yres;
-	l_hinc = (src_xres << 13) / dst_xres;
-	c_hinc = (src_xres << 13) / dst_xres;
+	l_vinc = ((src_yres << 13) + (dst_yres >> 1)) / dst_yres;
+	c_vinc = ((src_yres << 13) + (dst_yres >> 1)) / dst_yres;
+	l_hinc = ((src_xres << 13) + (dst_xres >> 1)) / dst_xres;
+	c_hinc = ((src_xres << 13) + (dst_xres >> 1)) / dst_xres;
 
 	if (pix_format == DRM_FORMAT_UYVY || pix_format == DRM_FORMAT_VYUY ||
 	    pix_format == DRM_FORMAT_YUYV || pix_format == DRM_FORMAT_YVYU) {
@@ -389,8 +389,8 @@ bool dcss_scaler_can_scale(struct dcss_soc *dcss, int ch_num,
 	const struct dcss_scaler_ratios *ratios_map = dcss_scaler_ratios;
 
 	/* Convert to fixed point. Easier to work with. */
-	vscale_fp = (src_yres << 13) / dst_yres;
-	hscale_fp = (src_xres << 13) / dst_xres;
+	vscale_fp = ((src_yres << 13) + (dst_yres >> 1)) / dst_yres;
+	hscale_fp = ((src_xres << 13) + (dst_xres >> 1)) / dst_xres;
 
 	if (scaler->ch_using_wrscl == -1 || scaler->ch_using_wrscl == ch_num)
 		ratios_map = dcss_scaler_wrscl_ratios;
-- 
1.7.9.5

